kotlinx-coroutines-core/common/src/JobSupport.kt

```diff
@@ -894,22 +926,23 @@ public open class JobSupport constructor(active: Boolean) : Job, ChildJob, Paren
             // If it just becomes cancelling --> must process cancelling notifications
             notifyRootCause = finishing.rootCause.takeIf { !wasCancelling }
         }
-        // process cancelling notification here -- it cancels all the children _before_ we start to to wait them (sic!!!)
+        // process cancelling notification here -- it cancels all the children _before_ we start to wait them (sic!!!)
         notifyRootCause?.let { notifyCancelling(list, it) }
         // now wait for children
-        val child = firstChild(state)
+        val child = list.nextChild()
         if (child != null && tryWaitForChild(finishing, child, proposedUpdate))
             return COMPLETING_WAITING_CHILDREN
+        list.close(LIST_CHILD_PERMISSION)
```

Comment:
Could you please elaborate on (and leave a comment in the code) about this sequence?

It's unclear why we first wait a child, then close a list, and then wait for the rest if the child was finished

Reply:
If you first close and then wait, then you'll forbid adding new children as soon as the job becomes completing, even if we have to wait arbitrarily long after that for the existing children to complete.